### PR TITLE
feat: redesign key Advanced assignments for authentic implementation (#440)

### PR DESCRIPTION
**Pull request by dhar174**
Closes #440`n`nRedesigns selected high-impact Advanced assignments with concrete scaffolding and authentic implementation pathways while preserving grading compatibility.

### CODE REVIEW

File: Advanced/assignments/Advanced_Day12_homework.ipynb [hunk 1]
```diff
@@ -54,14 +54,61 @@
...
+    "\n",
+    "passed = 0\n",
+    "for test_fn in [test_add_and_get_round_trip, test_title_is_trimmed, test_blank_title_raises]:\n",
+    "    test_fn()\n",
+    "    passed += 1\n",
+    "\n",
+    "print(\"Hour 45 | test runner: pytest\")\n",
+    "print(\"Hour 45 | fixture name: sample_tasks\")\n",
+    "print(f\"Hour 45 | passing unit tests: {passed}\")\n",
+    "print(\"Hour 45 | failed blank title case: caught\")\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day12_homework.ipynb:118]: Addressed in 89489ee: added a concrete `sample_tasks()` seeding helper and made the test execution block more pytest-style without changing the grading output contract.

File: Advanced/assignments/Advanced_Day12_homework.ipynb [hunk 2]
```diff
@@ -95,14 +142,44 @@
...
-    "# Goal: a tighter test suite with edge-case and integration confidence\n",
-    "# Best practice: Add edge cases and one lightweight integration test after unit tests pass.\n",
-    "# Pitfall to avoid: Chasing a coverage number without meaningful scenarios leaves blind spots.\n",
-    "# TODO: Write your solution here.\n",
-    "# TODO: Print labeled lines that match the Hour 46 contract in criteria.json.\n"
+    "payload_path = Path(\"data\") / \"task_snapshot.json\"\n",
+    "payload_path.parent.mkdir(parents=True, exist_ok=True)\n",
+    "json_payload = [task.__dict__ for task in service._tasks.values()]\n",
+    "payload_path.write_text(json.dumps(json_payload, indent=2), encoding=\"utf-8\")\n",
+    "loaded_payload = json.loads(payload_path.read_text(encoding=\"utf-8\"))\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day12_homework.ipynb]: Addressed in 89489ee: added a public `list_tasks()` method and removed direct reliance on `service._tasks` in the integration snapshot; JSON round-trip assertions now validate content explicitly.

File: Advanced/assignments/Advanced_Day12_homework.ipynb [hunk 3]
```diff
@@ -95,14 +142,44 @@
...
+    "\n",
+    "empty_search = service.search(\"nonexistent\")\n",
+    "\n",
+    "print(\"Hour 46 | coverage target: service layer\")\n",
+    "print(f\"Hour 46 | edge case tested: {duplicate_result}\")\n",
+    "print(\"Hour 46 | integration test: json round trip\")\n",
+    "print(\"Hour 46 | regression guard: search empty result\")\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day12_homework.ipynb:199]: Addressed in 89489ee: `empty_search` is now asserted (`assert empty_search == []`) so the regression guard is executable rather than label-only.

File: Advanced/assignments/Advanced_Day12_homework.ipynb [hunk 4]
```diff
@@ -136,14 +213,32 @@
...
+    "    \"    print('tracker app launched')\",\n",
+    "    \"\",\n",
+    "    \"if __name__ == '__main__':\",\n",
+    "    \"    main()\",\n",
+    "    \"\",\n",
+    "])\n",
+    "(tracker_pkg / \"__main__.py\").write_text(entrypoint_source, encoding=\"utf-8\")\n",
+    "\n",
+    "Path(\"requirements.txt\").write_text(\"pytest\\n\", encoding=\"utf-8\")\n",
+    "Path(\"README.md\").write_text(\"# Tracker App\\n\\n## Run\\npython -m tracker_app\\n\", encoding=\"utf-8\")\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day12_homework.ipynb]: Addressed in 89489ee: packaging artifacts now write under `packaging_artifacts/` rooted at the assignment/script directory (`__file__` when converted), so repo-root `README.md` / `requirements.txt` are no longer touched.

File: Advanced/assignments/Advanced_Day8_homework.ipynb [hunk 5]
```diff
@@ -56,14 +56,51 @@
...
+    "            (\"task-100\", \"Archive logs\", \"open\", 0),\n",
+    "            (\"task-101\", \"Prepare sprint report\", \"open\", 0),\n",
+    "            (\"task-102\", \"Close old tickets\", \"done\", 1),\n",
+    "        ],\n",
+    "    )\n",
+    "    conn.commit()\n",
+    "    return conn\n",
+    "\n",
+    "\n",
+    "db_file = Path(\"data\") / \"tracker.db\"\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day8_homework.ipynb]: Addressed in 89489ee: DB path is now rooted to assignment/script location (`assignment_root / data / tracker.db`) instead of relative CWD.

File: Advanced/assignments/Advanced_Day2_homework.ipynb [hunk 6]
```diff
@@ -53,14 +53,48 @@
...
+    "try:\n",
+    "    TaskRecord.from_payload({\"title\": \"   \", \"priority\": \"low\"})\n",
+    "except ValueError as exc:\n",
+    "    factory_error = str(exc)\n",
+    "\n",
+    "print(f\"Hour 5 | factory title: {factory_task.title}\")\n",
+    "print(f\"Hour 5 | factory priority: {factory_task.priority}\")\n",
+    "print(f\"Hour 5 | factory owner: {factory_task.owner}\")\n",
+    "print(f\"Hour 5 | factory tags: {', '.join(factory_task.tags)}\")\n",
+    "print(f\"Hour 5 | validation error: {factory_error}\")\n"
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day2_homework.ipynb:98]: Addressed in 89489ee: initialized `factory_error` before the try/except so execution is robust even if behavior changes.

File: Advanced/assignments/Advanced_Day12_homework.ipynb (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day12_homework.ipynb]: Addressed in 89489ee: packaging files are now emitted under assignment-scoped `packaging_artifacts/` to avoid root-file overwrite side effects.

File: Advanced/assignments/Advanced_Day12_homework.ipynb [hunk 7]
```diff
@@ -95,14 +142,44 @@
...
+    "\n",
+    "\n",
+    "service = AdvancedTaskService()\n",
+    "service.add_unique_task(\"task-1\", \"Plan sprint\")\n",
+    "service.add_unique_task(\"task-2\", \"Write release report\")\n",
+    "\n",
+    "try:\n",
+    "    service.add_unique_task(\"task-1\", \"Duplicate\")\n",
+    "except ValueError as exc:\n",
+    "    duplicate_result = str(exc)\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day12_homework.ipynb:183]: Addressed in 89489ee: initialized `duplicate_result` before the duplicate-ID check to avoid unbound-variable risk.

File: Advanced/assignments/Advanced_Day2_homework.ipynb [hunk 8]
```diff
@@ -53,14 +53,48 @@
...
+    "        \"title\": \"Refactor auth flow\",\n",
+    "        \"priority\": \"HIGH\",\n",
+    "        \"tags\": [\"API\", \"backend\"],\n",
+    "    }\n",
+    ")\n",
+    "\n",
+    "try:\n",
+    "    TaskRecord.from_payload({\"title\": \"   \", \"priority\": \"low\"})\n",
+    "except ValueError as exc:\n",
+    "    factory_error = str(exc)\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day2_homework.ipynb:92]: Addressed in 89489ee: initialized `factory_error` prior to exception handling to prevent a possible `NameError`.

File: Advanced/assignments/Advanced_Day12_homework.ipynb [hunk 9]
```diff
@@ -54,14 +54,61 @@
...
+    "    except ValueError as exc:\n",
+    "        assert str(exc) == \"title must not be blank\"\n",
+    "        return\n",
+    "    raise AssertionError(\"Expected ValueError for blank title\")\n",
+    "\n",
+    "\n",
+    "passed = 0\n",
+    "for test_fn in [test_add_and_get_round_trip, test_title_is_trimmed, test_blank_title_raises]:\n",
+    "    test_fn()\n",
+    "    passed += 1\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day12_homework.ipynb:113]: Addressed in 89489ee: test execution loop now catches assertion failures per test and continues collecting outcomes, rather than aborting at first failure.

File: Advanced/assignments/Advanced_Day12_homework.ipynb [hunk 10]
```diff
@@ -95,14 +142,44 @@
...
-    "# Hour 46 starter\n",
-    "# Focus: Coverage + edge cases + lightweight integration test\n",
-    "# Goal: a tighter test suite with edge-case and integration confidence\n",
-    "# Best practice: Add edge cases and one lightweight integration test after unit tests pass.\n",
-    "# Pitfall to avoid: Chasing a coverage number without meaningful scenarios leaves blind spots.\n",
-    "# TODO: Write your solution here.\n",
-    "# TODO: Print labeled lines that match the Hour 46 contract in criteria.json.\n"
+    "payload_path = Path(\"data\") / \"task_snapshot.json\"\n",
+    "payload_path.parent.mkdir(parents=True, exist_ok=True)\n",
+    "json_payload = [task.__dict__ for task in service._tasks.values()]\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day12_homework.ipynb]: Addressed in 89489ee: replaced dataclass `__dict__` serialization with `dataclasses.asdict()` and retained explicit round-trip assertions.

File: Advanced/assignments/Advanced_Day12_homework.ipynb [hunk 11]
```diff
@@ -136,14 +213,32 @@
...
-    "# Goal: a documented runnable delivery package for the capstone\n",
-    "# Best practice: Package the runnable app and its dependencies before optional build tooling.\n",
-    "# Pitfall to avoid: Spending all packaging time on an executable can crowd out clear run instructions.\n",
-    "# TODO: Write your solution here.\n",
-    "# TODO: Print labeled lines that match the Hour 47 contract in criteria.json.\n"
+    "# Hour 47 scaffold: packaging essentials for a runnable deliverable\n",
+    "tracker_pkg = Path(\"tracker_app\")\n",
+    "tracker_pkg.mkdir(parents=True, exist_ok=True)\n",
+    "(tracker_pkg / \"__init__.py\").write_text(\"\", encoding=\"utf-8\")\n",
+    "entrypoint_source = chr(10).join([\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day12_homework.ipynb]: Addressed in 89489ee: replaced `chr(10)` join with an idiomatic `"\n".join(...)` for readability.

File: Advanced/assignments/Advanced_Day12_homework.ipynb [hunk 12]
```diff
@@ -54,14 +54,61 @@
    "metadata": {},
    "outputs": [],
    "source": [
+    "# Hour 45 scaffold: behavior-focused unit tests for service logic\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day12_homework.ipynb:57]: Thanks — this is a pedagogical-scope concern. For this PR, I kept the existing graded contract shape and focused on correctness/safety defects. I agree this should be revisited as curriculum policy in a dedicated follow-up so assignment intent and grading strategy stay aligned.

File: Advanced/assignments/Advanced_Day2_homework.ipynb [hunk 13]
```diff
@@ -53,14 +53,48 @@
...
    "source": [
-    "\n",
-    "# Hour 5 starter\n",
-    "# Focus: Pattern: Factory (practical creation + validation)\n",
-    "# Goal: a validated factory that turns a raw payload into a safe tracker record\n",
-    "# Best practice: Use a classmethod factory to validate and normalize payloads before object creation.\n",
-    "# Pitfall to avoid: Duplicating validation logic in every caller leads to inconsistent objects.\n",
-    "# TODO: Write your solution here.\n",
-    "# TODO: Print labeled lines that match the Hour 5 contract in criteria.json.\n"
+    "# Hour 5 scaffold: Factory pattern with validation + normalization\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day2_homework.ipynb:56]: Thanks — same rationale as above: this comment is about assignment-template pedagogy rather than a correctness bug in this diff. I kept PR scope to correctness/safety fixes and recommend handling template-depth policy in a dedicated follow-up.

File: Advanced/assignments/Advanced_Day8_homework.ipynb [hunk 14]
```diff
@@ -56,14 +56,51 @@
...
    "source": [
-    "\n",
-    "# Hour 29 starter\n",
-    "# Focus: Deeper SQL practice (required) with ORM as an optional extension\n",
-    "# Goal: the deeper SQL query set plus an optional ORM extension note\n",
-    "# Best practice: Deliver the deeper SQL path first, then capture ORM as an optional extension note if you explored it.\n",
-    "# Pitfall to avoid: Adding ORM complexity without understanding the underlying query shape hides bugs.\n",
-    "# TODO: Write your solution here.\n",
-    "# TODO: Print labeled lines that match the Hour 29 contract in criteria.json.\n"
+    "# Hour 29 scaffold: deeper SQL queries with deterministic seed data\n",
```
[the remark this answers is not in the document]
Comment (dhar174) [reply Advanced/assignments/Advanced_Day8_homework.ipynb:59]: Thanks — same scope note: this is a curriculum policy concern. This PR now addresses concrete runtime and boundary defects; the “how much solution vs scaffold” question should be handled in a focused follow-up policy pass.
